File: top-k-frequent-elements/topKFrequent.cs
// https://leetcode.com/problems/top-k-frequent-elements/

public static int[] BucketSortSolution(int[] nums, int k)
{

    var buckets = new List<List<int>>();
    for (int x = 0; x <= nums.Length; x++) { 
        buckets.Add(new List<int>());
    }

    var frequencyCounter = new Dictionary<int, int>();
    foreach (int n in nums) {
        if (frequencyCounter.ContainsKey(n)) {
            frequencyCounter[n] += 1;
        } else {
            frequencyCounter.Add(n, 1);
        }
    }

    foreach (var (n, freq) in frequencyCounter) {
        buckets[freq].Add(n);	
    }

    var res = new int[k];
    var i = 0;
    var countDown = nums.Length; 
    while(countDown != 0) {
        foreach (var b in buckets[countDown]) {
            res[i] = b;
            i++;
            if (i == k) {
                return res;
            }
        }
        countDown--;
    }

    return res;
}

File: top-k-frequent-elements/topKFrequent.py
# https://leetcode.com/problems/top-k-frequent-elements/

def topKFrequent1(nums: list[int], k: int) -> list[int]:
    freq = dict()
    for n in nums:
        if n not in freq:
            freq[n] = 1
        else:
            freq[n] += 1
    sorted_dict = sorted(freq.items(), key=lambda x: x[1], reverse=True)

    res = [x for (x, y) in sorted_dict[0:k]]

    return res


def topKFrequent2(nums: list[int], k: int):
    freq = {}
    bucket = [[] for x in range(len(nums) + 1)]

    for n in nums:
        freq[n] = freq.get(n, 0) + 1

    for key, val in freq.items():
        bucket[val].append(key)

    res = []
    for i in range(len(bucket) - 1, 0, -1):
        for b in bucket[i]:
            res.append(b)
            if len(res) == k:
                return res


nums = [3, 3, 2, 3, 1, 2, 3, 2, 4]
k = 2
# res = topKFrequent(nums, k)
res = topKFrequentOptimal(nums, k)
print(res)
